backstage/backstage
# [Search] Refactor IndexBuilder as a stream pipeline

## 🌊  What / Why

Search indexing is very inefficient right now.  Each part of the process expects that the entire contents of an index is available in memory at once.  The proposal here is to alleviate this by reimagining each component of the search indexing process as a stream:

- `Collator` is just a `Readable` stream of `IndexableDocument` objects.
- `Decorator` is just a `Transform` stream operating on the same objects
- `SearchEngine::index()` is just a `Writable` stream, accepting the documents.

We can compose these streams like this:

```typescript
import { pipeline } from 'stream';
pipeline([collator, ...decorators, indexer], callback);
```

With a properly implemented stream pipeline, we should be able to get maximum performance for minimum memory.

### Making Things Stream-y
Properly implemented streams can be difficult and unintuitive.  To simplify for implementors (and for existing implementations), I've done the following:

- **Collators**: For existing collators, I've modified the `execute` methods to return `AsyncGenerator<IndexableDocument>`, and then converted those generators into readable streams via `Readable.from(collator.execute())`.  This should alleviate memory usage in the TechDocs collator (due to how it's already implemented), ~~but further optimizations would likely need to be made to the Catalog collator~~ and a minor update was made to the Catalog collator to take advantage of the (newly added) pagination parameters.  This seems like a reasonable pattern for future collators (which rely on data sources which themselves can't be streamed).
- **Indexers**: I created a simple `BatchSearchEngineIndexer` base class to simplify writable stream implementations for search engines.  The batch size is configurable and there are convenience hooks for async initialization and teardown.  Backpressure is handled by respecting the promise returned by the main `index` method. 
- **Decorators**: I created a simple `DecoratorBase` class to simplify transform stream implementations. Similar to the batch indexer, there are convenience hooks for async initialization and teardown.  It supports document modification, filtration, and fanning.  Backpressure is handled by respecting the promise returned by the main `decorate` method.

A cool benefit of this is that we open ourselves up to a world of stream-based sources for indexing (imagine very efficiently indexing gigabytes of newline-delimited JSON from cloud storage, for example) without having to write complex collator logic (we could instead rely on normal stream-based file or network calls).


### Backwards Compatibility

Proposal is to decouple the Catalog and TechDocs backend changes from the Search Changes.
- Adopters can upgrade (and continue to upgrade for a time) their TechDocs and/or Catalog backends without having to upgrade search.
- Because search is still in alpha (this is one of the latest items ahead of beta), let's take the opportunity to introduce a breaking change now.  If an adopter chooses to upgrade their search backend implementation, they will also have to upgrade TechDocs and Catalog backends.

---

### How to review

- Start with [types.ts](https://github.com/backstage/backstage/pull/7239/files#diff-a69818b67cf49b9857a2d85eaeacb51bcf61a5a6d7fc7518303488f417edd840)
- Check out the new [IndexBuilder implementation](https://github.com/backstage/backstage/pull/7239/files#diff-b866aa66f2c0c07a6b29ddb2053821cd63fe102b8033b171a1f62346f9a96423)
- Move on to implementations of new factories / search engine indexers (and base classes)

---

### ☑️ Todos
- [x] Validate the approach w/contributors
- [x] Iterate on / clean up implementations
- [x] Possibly create test utilities for dealing with streams
- [x] Provide an abstract class for decorators that simplifies implementation
- [x] Changesets, migration instructions
- [x] Backward compatibility decisions
---

Closes #6734 

#### :heavy_check_mark: Checklist
